src/pass/lift_if_then_else.cc

```diff
@@ -0,0 +1,422 @@
+      if (current_for.get() == for_node_list.back()) {
+        *ret = new_if_stmt;
+      }
+    });
+
+  return IRTransform(parent_for_stmt, nullptr, replace_target_for,
+                     {Expr("For")});
+}
+
+// Remove IfThenElse node from a For node.
+// A pair of For nodes will be generated.
+std::pair<Stmt, Stmt> RemoveIf(const Stmt& for_stmt, const Stmt& if_stmt) {
+  Stmt then_for;
+  Stmt else_for;
+  CHECK(if_stmt.as<IfThenElse>());
+
+  PackedFunc replace_then_case = PackedFunc(
+    [&](TVMArgs args, TVMRetValue *ret){
+      const NodeRef& node  = args[0];
+      if (node == if_stmt) {
+        *ret = node.as<IfThenElse>()->then_case;
+      }
+    });
+
+  PackedFunc replace_else_case = PackedFunc(
+    [&](TVMArgs args, TVMRetValue *ret){
+      const NodeRef& node  = args[0];
+      if (node == if_stmt) {
+        *ret = node.as<IfThenElse>()->else_case;
+      }
+    });
+
+  then_for = IRTransform(for_stmt, nullptr, replace_then_case,
+                         {Expr("IfThenElse")});
+  if (if_stmt.as<IfThenElse>()->else_case) {
+    else_for = IRTransform(for_stmt, nullptr, replace_else_case,
+                           {Expr("IfThenElse")});
+  }
+
+  return std::make_pair(then_for, else_for);
+}
+
+// Locate all For nodes and capture child IfThenElse nodes.
+void IfThenElseLHoist::SelectCandidates(const Stmt& stmt) {
+  PostOrderVisit(stmt, [&](const NodeRef& node){
+    const For* for_node = node.as<For>();
+    if (for_node) {
+      std::queue<Stmt> tracker;
+      tracker.push(for_node->body);
+      Stmt for_stmt = Downcast<Stmt, NodeRef>(node);
+      for2if_map_.insert({for_stmt.get(), std::vector<Stmt>()});
+      while (!tracker.empty()) {
+        Stmt head = tracker.front();
+        tracker.pop();
+        if (head->is_type<For>()) {
+          for (const auto& if_stmt : for2if_map_.at(head.get())) {
+            for2if_map_[for_stmt.get()].push_back(if_stmt);
+          }
+        } else if (head->is_type<AttrStmt>()) {
+          const AttrStmt* attr_node = head.as<AttrStmt>();
+          tracker.push(attr_node->body);
+        } else if (head->is_type<IfThenElse>()) {
+          for2if_map_[for_stmt.get()].push_back(head);
+          const IfThenElse* if_node = head.as<IfThenElse>();
+          tracker.push(if_node->then_case);
+          if (if_node->else_case) {
+            tracker.push(if_node->else_case);
+          }
+
+          // Record condition variables.
+          if (!cond_var_map_.count(head.get())) {
+            std::unordered_set<const Node*> new_var_set;
+            cond_var_map_.insert({head.get(), new_var_set});
+            PostOrderVisit(if_node->condition, [&](const NodeRef& cond_node) {
+              if (cond_node.as<Variable>()) {
+                cond_var_map_[head.get()].insert(cond_node.get());
+              }
+            });
+          }
+        } else {
+          continue;
+        }
+      }
+      ordered_for_list_.emplace_back(Downcast<Stmt, NodeRef>(node));
```

Comment:
Trying to understand. Most of the other bookkeeping has been done for `Node*` type, but this one is for `Stmt`. Any reason we want to do that?

Reply:
When generating if2for_map_, we need to push a Stmt object into each value vector. Here we directly store Stmt object so that we don't need to use For::make to build a Stmt again.